implementation verify.go

fehlmach opened the pull request:
implemented verify.go

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/netsec-ethz/rains/6)
<!-- Reviewable:end -->

klausman commented:
First pass.

---

Reviewed 14 of 14 files at r1.
Review status: all files reviewed at latest revision, 34 unresolved discussions.

---

*[rainsd/inbox.go, line 121 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgEHl6FYpFSnEwTGWda:-KgEHl6FYpFSnEwTGWdb:b-dfmdin) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/inbox.go#L121)):*
> ```Go
> 	msg, err := CreateNotificationMsg(token, notificationType, "")
> 	if err != nil {
> 		log.Warn("Cannot send notification error due to parser error")
> ```

Log the error here, if possible/useful.

---

*[rainsd/inbox.go, line 156 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgEIHi--wV8lRQCLeoD:-KgEIHi0aRT4TNYiX5NJ:bdgycaf) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/inbox.go#L156)):*
> ```Go
> 		notificationChannel <- MsgSectionSender{Sender: sender, Msg: msg, Token: tok}
> 	} else {
> 		log.Warn("Token not in active token cache, drop message", "Token", tok)
> ```

We usually use all-lowercase identifiers in the annotated logs (`Token` -> `token`)

---

*[rainsd/inbox.go, line 172 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgEISxX2cJeHvaXH6WG:-KgEISxX2cJeHvaXH6WH:boy976s) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/inbox.go#L172)):*
> ```Go
> 		go workNotification()
> 	}
> 	for i := 0; i < int(2); i++ {
> ```

why the `int()` cast?

---

*[rainsd/inbox.go, line 210 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgEIixku-fk9c750j3a:-KgEIixku-fk9c750j3b:buezvr9) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/inbox.go#L210)):*
> ```Go
> 		select {
> 		case msg := <-notificationChannel:
> 			Notify(msg)
> ```

What is the rationale for making `verify()` private and `Notify()` public?

---

*[rainsd/notification.go, line 20 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgEJ21l5zialW-iHRIb:-KgEJ21l5zialW-iHRIc:b-xl5r3g) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/notification.go#L20)):*
> ```Go
> 		//store in a map key <dest> value <capabilities>. Before a message is parsed to CBOR format check if it must include capabilities.
> 	case rainslib.BadMessage:
> 		notifLog.Error("Sent msg was malformated")
> ```

malformed*

---

*[rainsd/serverInfo.go, line 21 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgEJFYAK_iPZcGOVA2p:-KgEJFYAK_iPZcGOVA2q:bnbgaxl) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/serverInfo.go#L21)):*
> ```Go
> type rainsdConfig struct {
> 	//switchboard
> 	ServerIPAddr    string
> ```

Could this be a `net.IP`?

---

*[rainsd/serverInfo.go, line 34 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgEJPOqxpXdLmJ558RW:-KgEJPOqxpXdLmJ558RX:b-r5gp05) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/serverInfo.go#L34)):*
> ```Go
> 	NormalBufferSize       uint
> 	NotificationBufferSize uint
> 	PrioWorkerSize         uint
> ```

Minor nit: I think this is a `Count`, rather than a `Size`

---

*[rainsd/serverInfo.go, line 63 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgEJcqeOcOqKSrdsna8:-KgEJcqeOcOqKSrdsna9:bwsxs2t) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/serverInfo.go#L63)):*
> ```Go
> 
> //ConnInfo contains address information about one actor of a connection of the declared type
> //type 1 contains IPAddr and Port information
> ```

Is this `1` an editing remnant?

---

*[rainsd/serverInterfaceImpl.go, line 38 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgEmQFwjwHjRF9jLmzJ:-KgEmQFwjwHjRF9jLmzK:b-lppjts) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/serverInterfaceImpl.go#L38)):*
> ```Go
> }
> 
> //TODO CFE replace this with an own implementation
> ```

Why do we need to replace this?

---

*[rainsd/serverUtil.go, line 99 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgEmx4kDd_F5O78LfJ9:-KgEmx4kDd_F5O78LfJA:b69yftm) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/serverUtil.go#L99)):*
> ```Go
> 		log.Warn("Could not cast key to ed25519.PrivateKey", "privateKey", privateKey)
> 	case rainslib.Ed448:
> 		log.Warn("Not yet Supported!")
> ```

Add `Ed448` to the log message.

---

*[rainsd/serverUtil.go, line 103 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgEn4AlDvtDzIDfnnmr:-KgEn4AlDvtDzIDfnnms:bq9f6ma) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/serverUtil.go#L103)):*
> ```Go
> 		if pkey, ok := privateKey.(*ecdsa.PrivateKey); ok {
> 			//TODO CFE or use sha256?
> 			hash := sha256.Sum256(data)
> ```

This looks like the code and comment disagree.

---

*[rainsd/serverUtil.go, line 136 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgEnIi9mu3XgcVlNxo9:-KgEnIi9mu3XgcVlNxoA:b2nreh6) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/serverUtil.go#L136)):*
> ```Go
> 		log.Warn("Could not cast key to ed25519.PublicKey", "publicKey", publicKey)
> 	case rainslib.Ed448:
> 		log.Warn("Not yet Supported!")
> ```

Same as above.

---

*[rainsd/switchboard.go, line 100 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgEoKFKPnuseWoe039e:-KgEoKFKPnuseWoe039f:b2skl8) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/switchboard.go#L100)):*
> ```Go
> 		return fmt.Sprintf("%s_%d_%s_%d", client.IPAddr, client.Port, server.IPAddr, server.Port)
> 	default:
> 		log.Warn("le(): No matching type found for client ConnInfo")
> ```

I think there is something missing in front of the `le()`

---

*[rainsd/verify.go, line 73 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgEsNUkB66vsQ9ox190:-KgEsNUkB66vsQ9ox191:bpbv78) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/verify.go#L73)):*
> ```Go
> 	case *rainslib.AssertionSection, *rainslib.ShardSection, *rainslib.ZoneSection:
> 		if !containedAssertionsAndShardsValid(msgSender) {
> 			return
> ```

This should probably log that the verification failed (unless the function called already does so, then maybe add a comment here clarifying that).

---

*[rainsd/verify.go, line 76 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgEsblTGydLQdfU11z2:-KgEsblTGydLQdfU11z3:bhq2r5g) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/verify.go#L76)):*
> ```Go
> 		}
> 		if verifySignatures(section.(rainslib.MessageSectionWithSig), msgSender.Sender, false, nil) {
> 			Assert(section.(rainslib.MessageSectionWithSig))
> ```

Does `Assert()` return anything useful at all?

---

*[rainsd/verify.go, line 108 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgEv4aGxbyyV5PVgeB3:-KgEv4aHKjAMrFg8UdJE:bm2qnm0) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/verify.go#L108)):*
> ```Go
> //containedAssertionsAndShardsValid compares the context and the subject zone of the outer message section with the contained message bodies.
> //If they differ, an inconsistency notification msg is sent to the sender and false is returned
> func containedAssertionsAndShardsValid(msgSender MsgSectionSender) bool {
> ```

I find the logic in this function a bit hard to follow. Can you annotate the switch/case branches with comments to make it clearer?

---

*[rainsd/verify.go, line 170 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgEvQg8rdNkE039Atgc:-KgEvQg8rdNkE039Atgd:blz4lyi) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/verify.go#L170)):*
> ```Go
> 			log.Warn("Unsupported keyspace", "KeySpaceID", sig.KeySpace)
> 		}
> 		cacheKey := PendingSignatureCacheKey{KeySpace: strconv.Itoa(int(sig.KeySpace)), Context: section.GetContext(), SubjectZone: section.GetSubjectZone()}
> ```

Change this to an indented multi-line struct for readability, i.e.:
```
PendingSignatureCacheKey{
    KeySpace: strconv.Itoa(int(sig.KeySpace)),
    ...
}
```

---

*[rainsd/verify.go, line 198 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgExRAQ2cjhXgy93I1g:-KgExRAQ2cjhXgy93I1h:b-wod4vq) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/verify.go#L198)):*
> ```Go
> //publicKeysPresent returns true if all public keys are in the cache together with a map of keys and missingKeys
> func publicKeysPresent(keyIDs map[PendingSignatureCacheKey]string) (map[string]rainslib.PublicKey, map[PendingSignatureCacheKey]bool, bool) {
> 	allKeysPresent := true
> ```

drop this var and change the return below to `return ..., len(missingKeys) == 0` It's more defensive (harder to get the logic wrong).

---

*[rainsd/verify.go, line 209 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgExGXaVhOg50tlBZ3K:-KgExGXaVhOg50tlBZ3L:b-zcyhiw) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/verify.go#L209)):*
> ```Go
> 				missingKeys[keyID] = true
> 			} else {
> 				log.Info("Corresponding Public key in cash.", "CacheKey", keyID, "PublicKey", key)
> ```

s/cash/cache/

---

*[rainsd/verify.go, line 229 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgExvUdSdUCLRn7BtHR:-KgExvUdSdUCLRn7BtHS:blpx1ke) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/verify.go#L229)):*
> ```Go
> //handleMissingPublicKey adds or updates the current msg section to pendingSignatures and sends a delegation query for the missing key
> func handleMissingPublicKey(cacheKey PendingSignatureCacheKey, sender ConnInfo, section rainslib.MessageSectionWithSig, value *PendingSignatureCacheValue) {
> 	//TODO CFE How large should we set the expiration time of the query
> ```

I think it should be configurable and have a sane default (which we will likely have to figure out experimentally).

---

*[rainsd/verify.go, line 255 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgEyBHM2bBG70ajvFne:-KgEyBHM2bBG70ajvFnf:b-2jk7k2) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/verify.go#L255)):*
> ```Go
> func sendDelegationQuery(cacheKey PendingSignatureCacheKey, expTime int64, sender ConnInfo) {
> 	token := GenerateToken()
> 	querySection := rainslib.QuerySection{Context: cacheKey.Context, SubjectName: cacheKey.SubjectZone, Expires: int(expTime), Token: token, Types: rainslib.Delegation}
> ```

Turn this into a multi-line literal for easier reading.

---

*[rainslib/datamodel.go, line 158 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgFDbeNVRqdXnylXvQm:-KgFDbeNVRqdXnylXvQn:bf4gtur) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainslib/datamodel.go#L158)):*
> ```Go
> 			section.DeleteAllSigs()
> 		case *ShardSection:
> 			section.DeleteAllSigs()
> ```

These two cases look identical, can they be merged?

---

*[rainslib/datamodel.go, line 185 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgF0WdW2Yd2vWfWIrEB:-KgF0WdW2Yd2vWfWIrEC:by29x0q) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainslib/datamodel.go#L185)):*
> ```Go
> 			stub.Content = append(stub.Content, section.CreateStub())
> 		case *ShardSection:
> 			stub.Content = append(stub.Content, section.CreateStub())
> ```

This case and the above look identical. Can they be merged?

---

*[rainslib/datamodel.go, line 207 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgF0uUS1YpoHNDT-g_N:-KgF0uUTKB7uIXr3mtJy:b-9vvn8y) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainslib/datamodel.go#L207)):*
> ```Go
> }
> 
> type QueryOptions int
> ```

I think this should be singular, just like `ObjectType` below.

---

*[rainslib/datamodel.go, line 210 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgF0eDBLXKKBqcYrDRx:-KgF0eDBLXKKBqcYrDRy:b-1gzwd3) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainslib/datamodel.go#L210)):*
> ```Go
> 
> const (
> 	MinE2ELatency            QueryOptions = 1
> ```

Couldn't this be done with an `InvalidOption QueryOption iota` and then dropping the explicit ints? That way, an object that has been initialized by Go, but not filled in will automatically be invalid and have a corresponding name (same for `ObjectTypes` below). 

---

*[rainslib/datamodel.go, line 317 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgF1HYkgJgwwNo7nCIM:-KgF1HYkgJgwwNo7nCIN:bpcshoi) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainslib/datamodel.go#L317)):*
> ```Go
> type SignatureAlgorithmType int
> 
> const (
> ```

Same `iota` scheme as above.

---

*[rainslib/datamodel.go, line 328 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgF1F68Ho5W2H8_ZIEm:-KgF1F69DolpuDF4Ynsy:b58qgwq) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainslib/datamodel.go#L328)):*
> ```Go
> 
> const (
> 	NoHashAlgo HashAlgorithmType = 0
> ```

This could use `iota` directly.

---

*[server/server, line 0 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgF1orTwi2nxbld8lUF:-KgF1orTwi2nxbld8lUG:b3lajiu) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/server/server#L0)):*
What is this file's purpose?

---

*[server/config/server.conf, line 5 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgFDmUgwPwMCuIqSkCj:-KgFDmUgwPwMCuIqSkCk:bj9nxga) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/server/config/server.conf#L5)):*
> ```
> "ServerPort": 5022,
> "MaxConnections": 1000,
> "KeepAlivePeriod": 60000000000,
> ```

Expressions like this should have units, e.g. `KeepAlivePeriodMicroSeconds` (or `...Micros`)

---

*[server/config/server.conf, line 7 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgFDzxBKUxX-gZ0KWN_:-KgFDzxCaGAzGycptFbw:b-oh9bk) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/server/config/server.conf#L7)):*
> ```
> "KeepAlivePeriod": 60000000000,
> "TCPTimeout": 300000000000,
> "MaxMsgLength": 20000,
> ```

similar to the above, a unit would be useful if it's feasible: bytes? If it's just a count (like "X sub-message"), it's less useful, but could probably use  a count. The same goes for the BufferSize items below.

---

*[utils/parser/rainsMsgParser.go, line 18 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgF2M6_j15CtuJqcQPO:-KgF2M6_j15CtuJqcQPP:b-1l2kwr) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/utils/parser/rainsMsgParser.go#L18)):*
> ```Go
> //ParseByteSlice parses the byte slice to a RainsMessage according to the following format:
> //It is ASSUMED that signature data does not contain the '[' char.
> //RainsMessage format: <token>[MessageSection:::::...:::::MessageSection][signatures*]:cap:<capabilities
> ```

What do the runs of colons represent?

---

*[utils/parser/rainsMsgParser.go, line 295 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgF3A1ZCKbQfXFK7VpO:-KgF3A1ZCKbQfXFK7VpP:b-un0mcd) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/utils/parser/rainsMsgParser.go#L295)):*
> ```Go
> 		return &rainslib.AssertionSection{}, err
> 	}
> 	assertionSection := rainslib.AssertionSection{Context: context, SubjectZone: subjectZone, SubjectName: msg[sn+4 : objBegin], Content: objects, Signatures: signatures}
> ```

Make this a multi-line literal.

---

*[utils/parser/rainsMsgParser.go, line 476 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgFD27AhVlAlc3MkpNc:-KgFD27AhVlAlc3MkpNd:b-ab6z2g) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/utils/parser/rainsMsgParser.go#L476)):*
> ```Go
> 			return []rainslib.Signature{}, errors.New("signature's cipher malformated")
> 		}
> 		signature := rainslib.Signature{Algorithm: rainslib.SignatureAlgorithmType(algoType), Data: []byte(sig[sd+4 : len(sig)]), ValidSince: validSince, ValidUntil: validUntil,
> ```

Make this literal multi-lined (one element per line).

---

*[utils/parser/rainsMsgParser.go, line 516 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgFDMGb2mBS3FnYLVsA:-KgFDMGb2mBS3FnYLVsB:b7jl0ox) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/utils/parser/rainsMsgParser.go#L516)):*
> ```Go
> 		opts = append(opts, rainslib.QueryOptions(val))
> 	}
> 	return &rainslib.QuerySection{Token: token, Expires: expires, Context: msg[cn+4 : sn], SubjectName: msg[sn+4 : ot], Types: rainslib.ObjectType(objType), Options: opts}, nil
> ```

Again, multiline literal.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-:-KgFEKG88JUyLxDdyoNK:b-xepgfs)*
<!-- Sent from Reviewable.io -->

fehlmach commented:
Review status: 2 of 14 files reviewed at latest revision, 34 unresolved discussions.

---

*[rainsd/inbox.go, line 121 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgEHl6FYpFSnEwTGWda:-KgJuhkubdVuLaJLUQqG:b-896fix) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/inbox.go#L121)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary><blockquote>

Log the error here, if possible/useful.
</blockquote></details>

Done.

---

*[rainsd/inbox.go, line 156 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgEIHi--wV8lRQCLeoD:-KgJdINMg-Febn10zRPR:b2b3mz7) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/inbox.go#L156)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary><blockquote>

We usually use all-lowercase identifiers in the annotated logs (`Token` -> `token`)
</blockquote></details>

Done. changed it, hope i got all of them.

---

*[rainsd/inbox.go, line 172 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgEISxX2cJeHvaXH6WG:-KgJdW6aawhUGGGeDnq0:b-1m82qf) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/inbox.go#L172)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary><blockquote>

why the `int()` cast?
</blockquote></details>

Done. copy paste error

---

*[rainsd/inbox.go, line 210 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgEIixku-fk9c750j3a:-KgJdpePOW4wfnp9MEZJ:bnfk0xg) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/inbox.go#L210)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary><blockquote>

What is the rationale for making `verify()` private and `Notify()` public?
</blockquote></details>

Done. Notify should also be private I forgot to change it. Only Listen() in the switchboard needs to be public.

---

*[rainsd/notification.go, line 20 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgEJ21l5zialW-iHRIb:-KgJfHxUVzv7i4lav-mZ:b-896fix) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/notification.go#L20)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary><blockquote>

malformed*
</blockquote></details>

Done.

---

*[rainsd/serverInfo.go, line 21 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgEJFYAK_iPZcGOVA2p:-KgJj05IfMDc3OR9Qn4C:b-896fix) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/serverInfo.go#L21)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary><blockquote>

Could this be a `net.IP`?
</blockquote></details>

Done.

---

*[rainsd/serverInfo.go, line 34 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgEJPOqxpXdLmJ558RW:-KgJl08O-ok4G9-HIySu:b-896fix) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/serverInfo.go#L34)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary><blockquote>

Minor nit: I think this is a `Count`, rather than a `Size`
</blockquote></details>

Done.

---

*[rainsd/serverInfo.go, line 63 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgEJcqeOcOqKSrdsna8:-KgJlHRNZG6COFXb1tyW:b-pyyx2t) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/serverInfo.go#L63)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary><blockquote>

Is this `1` an editing remnant?
</blockquote></details>

Done. Yep

---

*[rainsd/serverInterfaceImpl.go, line 38 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgEmQFwjwHjRF9jLmzJ:-KgJx6cNlkZQ1a2WhQI6:bftiejp) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/serverInterfaceImpl.go#L38)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary><blockquote>

Why do we need to replace this?
</blockquote></details>

Because we also need to be able to match only parts of the key

---

*[rainsd/serverUtil.go, line 99 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgEmx4kDd_F5O78LfJ9:-KgJlX8PNKXOggb8McgG:b-896fix) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/serverUtil.go#L99)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary><blockquote>

Add `Ed448` to the log message.
</blockquote></details>

Done.

---

*[rainsd/serverUtil.go, line 103 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgEn4AlDvtDzIDfnnmr:-KgJlikDt9epkfYMy-pK:b-yq9850) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/serverUtil.go#L103)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary><blockquote>

This looks like the code and comment disagree.
</blockquote></details>

Done. Yes I ask Brian and I fixed it, but forgot the remove the comment

---

*[rainsd/serverUtil.go, line 136 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgEnIi9mu3XgcVlNxo9:-KgJlubh3PqL70BzZRGr:b-896fix) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/serverUtil.go#L136)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary><blockquote>

Same as above.
</blockquote></details>

Done.

---

*[rainsd/switchboard.go, line 100 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgEoKFKPnuseWoe039e:-KgJm9GytoWrqhxFopsz:b-896fix) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/switchboard.go#L100)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary><blockquote>

I think there is something missing in front of the `le()`
</blockquote></details>

Done.

---

*[rainsd/verify.go, line 73 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgEsNUkB66vsQ9ox190:-KgJnEETDyK12BFsWVpd:b-896fix) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/verify.go#L73)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary><blockquote>

This should probably log that the verification failed (unless the function called already does so, then maybe add a comment here clarifying that).
</blockquote></details>

Done.

---

*[rainsd/verify.go, line 76 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgEsblTGydLQdfU11z2:-KgJnLbsnJLg3h8s5ccl:b-yw4utg) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/verify.go#L76)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary><blockquote>

Does `Assert()` return anything useful at all?
</blockquote></details>

No, we just pass the section along and the engine should decide what to do with it. 

---

*[rainsd/verify.go, line 108 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgEv4aGxbyyV5PVgeB3:-KgJnjj1BEHYIb5J4Hxg:b-896fix) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/verify.go#L108)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary><blockquote>

I find the logic in this function a bit hard to follow. Can you annotate the switch/case branches with comments to make it clearer?
</blockquote></details>

Done.

---

*[rainsd/verify.go, line 170 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgEvQg8rdNkE039Atgc:-KgJoiOtM7BQ90iUs7Is:b-896fix) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/verify.go#L170)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary><blockquote>

Change this to an indented multi-line struct for readability, i.e.:
```
PendingSignatureCacheKey{
    KeySpace: strconv.Itoa(int(sig.KeySpace)),
    ...
}
```
</blockquote></details>

Done.

---

*[rainsd/verify.go, line 198 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgExRAQ2cjhXgy93I1g:-KgJpOtb4iErS6RMN-me:b-896fix) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/verify.go#L198)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary><blockquote>

drop this var and change the return below to `return ..., len(missingKeys) == 0` It's more defensive (harder to get the logic wrong).
</blockquote></details>

Done.

---

*[rainsd/verify.go, line 209 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgExGXaVhOg50tlBZ3K:-KgJpBOTqRtCJLd137Ps:b-896fix) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/verify.go#L209)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary><blockquote>

s/cash/cache/
</blockquote></details>

Done.

---

*[rainsd/verify.go, line 229 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgExvUdSdUCLRn7BtHR:-KgJps1VKsFMX6j5WIrW:b-896fix) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/verify.go#L229)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary><blockquote>

I think it should be configurable and have a sane default (which we will likely have to figure out experimentally).
</blockquote></details>

Done.

---

*[rainsd/verify.go, line 255 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgEyBHM2bBG70ajvFne:-KgJq-tcGRt5FopN0kBg:b-896fix) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainsd/verify.go#L255)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary><blockquote>

Turn this into a multi-line literal for easier reading.
</blockquote></details>

Done.

---

*[rainslib/datamodel.go, line 158 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgFDbeNVRqdXnylXvQm:-KgJqSlq8Y7yM5mDYRyf:b-896fix) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainslib/datamodel.go#L158)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary><blockquote>

These two cases look identical, can they be merged?
</blockquote></details>

Done.

---

*[rainslib/datamodel.go, line 185 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgF0WdW2Yd2vWfWIrEB:-KgJqNXSYDaaOeS59NpL:b-896fix) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainslib/datamodel.go#L185)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary><blockquote>

This case and the above look identical. Can they be merged?
</blockquote></details>

Done.

---

*[rainslib/datamodel.go, line 207 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgF0uUS1YpoHNDT-g_N:-KgJqdMfLJTajPpyMfdS:b-896fix) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainslib/datamodel.go#L207)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary><blockquote>

I think this should be singular, just like `ObjectType` below.
</blockquote></details>

Done.

---

*[rainslib/datamodel.go, line 210 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgF0eDBLXKKBqcYrDRx:-KgJr-953H28cgeqgRLG:b-blq3mr) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainslib/datamodel.go#L210)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary><blockquote>

Couldn't this be done with an `InvalidOption QueryOption iota` and then dropping the explicit ints? That way, an object that has been initialized by Go, but not filled in will automatically be invalid and have a corresponding name (same for `ObjectTypes` below). 
</blockquote></details>

The thing is that these int values  are specified with this value in the protocol specification. If we have it this way, the mapping is fixed between number and name and if someone is changing the order of these elements it has no effect. With an iota it would be very hard to detect when two names would have been swapped and thus, have a different meaning.  

---

*[rainslib/datamodel.go, line 317 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgF1HYkgJgwwNo7nCIM:-KgJrpVMn1_64QbH-1vg:byr4ffv) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainslib/datamodel.go#L317)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary><blockquote>

Same `iota` scheme as above.
</blockquote></details>

see above

---

*[rainslib/datamodel.go, line 328 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgF1F68Ho5W2H8_ZIEm:-KgJrtSyw4AKD3friiRo:byr4ffv) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainslib/datamodel.go#L328)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary><blockquote>

This could use `iota` directly.
</blockquote></details>

see above

---

*[server/config/server.conf, line 5 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgFDmUgwPwMCuIqSkCj:-KgJsm6HeoKqVDFnhZ-R:b-896fix) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/server/config/server.conf#L5)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary><blockquote>

Expressions like this should have units, e.g. `KeepAlivePeriodMicroSeconds` (or `...Micros`)
</blockquote></details>

Done.

---

*[server/config/server.conf, line 7 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgFDzxBKUxX-gZ0KWN_:-KgJsmO5thJLuNV2U0jn:b-896fix) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/server/config/server.conf#L7)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary><blockquote>

similar to the above, a unit would be useful if it's feasible: bytes? If it's just a count (like "X sub-message"), it's less useful, but could probably use  a count. The same goes for the BufferSize items below.
</blockquote></details>

Done.

---

*[utils/parser/rainsMsgParser.go, line 18 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgF2M6_j15CtuJqcQPO:-KgJtOBbEHRGwCKspIqN:b4i1m8w) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/utils/parser/rainsMsgParser.go#L18)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary><blockquote>

What do the runs of colons represent?
</blockquote></details>

a delimiter between MessageSections

---

*[utils/parser/rainsMsgParser.go, line 295 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgF3A1ZCKbQfXFK7VpO:-KgJuCO60WobDkb3Ev6j:b-896fix) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/utils/parser/rainsMsgParser.go#L295)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary><blockquote>

Make this a multi-line literal.
</blockquote></details>

Done.

---

*[utils/parser/rainsMsgParser.go, line 476 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgFD27AhVlAlc3MkpNc:-KgJuCv-k1LE1XYR4dS3:b-896fix) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/utils/parser/rainsMsgParser.go#L476)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary><blockquote>

Make this literal multi-lined (one element per line).
</blockquote></details>

Done.

---

*[utils/parser/rainsMsgParser.go, line 516 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgFDMGb2mBS3FnYLVsA:-KgJuD_TpzFBiuFTm3H4:b-896fix) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/utils/parser/rainsMsgParser.go#L516)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary><blockquote>

Again, multiline literal.
</blockquote></details>

Done.

---

*[server/server, line  at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgF1orTwi2nxbld8lUF:-KgJryT_Nl_kKl9COOI_:bsaud4x) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/server/server#L)):*
<details><summary><i>Previously, klausman (Tobias Klausmann) wrote…</i></summary><blockquote>

What is this file's purpose?
</blockquote></details>

I use it to start the server, it should not be in the repo.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/rains/6)*
<!-- Sent from Reviewable.io -->

klausman commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> from me, on to @britram 

---

Reviewed 13 of 13 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[rainslib/datamodel.go, line 210 at r1](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgF0eDBLXKKBqcYrDRx:-KgKDECF22_HcugNVwL8:b-vawk5r) ([raw file](https://github.com/netsec-ethz/rains/blob/d4b488229e5727fdfa2b265b07da94c09aca186c/rainslib/datamodel.go#L210)):*
<details><summary><i>Previously, fehlmach wrote…</i></summary><blockquote>

The thing is that these int values  are specified with this value in the protocol specification. If we have it this way, the mapping is fixed between number and name and if someone is changing the order of these elements it has no effect. With an iota it would be very hard to detect when two names would have been swapped and thus, have a different meaning.  
</blockquote></details>

Okay, that's a good enough reason to make it static/explicit.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-:-KgKDm8m3FqA1a1N7tZ1:b-xi4862)*
<!-- Sent from Reviewable.io -->

britram commented:
Reviewed 2 of 14 files at r1, 11 of 13 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[rainsd/engine.go, line 20 at r2](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgLqflV8eIwEkcr9BpC:-KgLqflV8eIwEkcr9BpD:bwjg9g0) ([raw file](https://github.com/netsec-ethz/rains/blob/6d4b139b5f1e9d0b94f36be605b10b82ce4876bf/rainsd/engine.go#L20)):*
> ```Go
> 
> //pendingQueries contains a mapping from all self issued pending queries to the set of message bodies waiting for it.
> //key: <keyspace><context><subjectzone> value: <msgSection><deadline>
> ```

FYI: self issued queries will *never* occur in any other than the RAINS keyspace. keyspace is only relevant for rainsd in that non-RAINS keyspaces mean
 "rainsd can safely ignore this signature" 

---

*[rainsd/inbox.go, line 162 at r2](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgLukxm2pW5fCsNjlv3:-KgLukxm2pW5fCsNjlv4:b-kdwdi5) ([raw file](https://github.com/netsec-ethz/rains/blob/6d4b139b5f1e9d0b94f36be605b10b82ce4876bf/rainsd/inbox.go#L162)):*
> ```Go
> //createWorker creates go routines which process messages from the prioChannel and normalChannel.
> //number of go routines per queue are loaded from the config
> func createWorker() {
> ```

BTW not clear to me that the worker pattern is the right way to go... but let's see how well this works.

---

*[rainsd/inbox.go, line 188 at r2](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgLwdpRLC5746CgB50Z:-KgLwdpRLC5746CgB50_:b55d2di) ([raw file](https://github.com/netsec-ethz/rains/blob/6d4b139b5f1e9d0b94f36be605b10b82ce4876bf/rainsd/inbox.go#L188)):*
> ```Go
> 		default:
> 			//TODO CFE add to config?
> 			time.Sleep(50 * time.Millisecond)
> ```

why is this here?

---

*[rainsd/inbox.go, line 201 at r2](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgLwiRCvnLVcsrZACVB:-KgLwiRCvnLVcsrZACVC:b55d2di) ([raw file](https://github.com/netsec-ethz/rains/blob/6d4b139b5f1e9d0b94f36be605b10b82ce4876bf/rainsd/inbox.go#L201)):*
> ```Go
> 		default:
> 			//TODO CFE add to config?
> 			time.Sleep(50 * time.Millisecond)
> ```

why is this here?

---

*[rainsd/inbox.go, line 214 at r2](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgLwkeV134tdt7TEGNV:-KgLwkeW9pRdcOE7Zhas:b55d2di) ([raw file](https://github.com/netsec-ethz/rains/blob/6d4b139b5f1e9d0b94f36be605b10b82ce4876bf/rainsd/inbox.go#L214)):*
> ```Go
> 		default:
> 			//TODO CFE add to config?
> 			time.Sleep(50 * time.Millisecond)
> ```

why is this here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/rains/6)*
<!-- Sent from Reviewable.io -->

fehlmach commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[rainsd/engine.go, line 20 at r2](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgLqflV8eIwEkcr9BpC:-KgNqJUKHN2_L2qtPBer:b-896fix) ([raw file](https://github.com/netsec-ethz/rains/blob/6d4b139b5f1e9d0b94f36be605b10b82ce4876bf/rainsd/engine.go#L20)):*
<details><summary><i>Previously, britram (Brian Trammell) wrote…</i></summary><blockquote>

FYI: self issued queries will *never* occur in any other than the RAINS keyspace. keyspace is only relevant for rainsd in that non-RAINS keyspaces mean
 "rainsd can safely ignore this signature" 

</blockquote></details>

Done.

---

*[rainsd/inbox.go, line 162 at r2](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgLukxm2pW5fCsNjlv3:-KgNohwIIKsC1SdO59OM:b3xev09) ([raw file](https://github.com/netsec-ethz/rains/blob/6d4b139b5f1e9d0b94f36be605b10b82ce4876bf/rainsd/inbox.go#L162)):*
<details><summary><i>Previously, britram (Brian Trammell) wrote…</i></summary><blockquote>

BTW not clear to me that the worker pattern is the right way to go... but let's see how well this works.
</blockquote></details>

Yes, but I think we need some mechanism to have an upper limit of running go routines.  Otherwise if too many assertions/queries are coming in, we might run out of memory and the server crashes. 

---

*[rainsd/inbox.go, line 188 at r2](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgLwdpRLC5746CgB50Z:-KgNpN7Ko7ns2-3mq0wX:b-roxoiu) ([raw file](https://github.com/netsec-ethz/rains/blob/6d4b139b5f1e9d0b94f36be605b10b82ce4876bf/rainsd/inbox.go#L188)):*
<details><summary><i>Previously, britram (Brian Trammell) wrote…</i></summary><blockquote>

why is this here?
</blockquote></details>

To avoid spinning all the time. If there are no messages to process available then wait some time before looking again.  But we can remove it. 

---

*[rainsd/inbox.go, line 201 at r2](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgLwiRCvnLVcsrZACVB:-KgNqKe0Mra20eg1DvVc:b-896fix) ([raw file](https://github.com/netsec-ethz/rains/blob/6d4b139b5f1e9d0b94f36be605b10b82ce4876bf/rainsd/inbox.go#L201)):*
<details><summary><i>Previously, britram (Brian Trammell) wrote…</i></summary><blockquote>

why is this here?
</blockquote></details>

Done.

---

*[rainsd/inbox.go, line 214 at r2](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgLwkeV134tdt7TEGNV:-KgNqL-1eYGaufsdNHAA:b-896fix) ([raw file](https://github.com/netsec-ethz/rains/blob/6d4b139b5f1e9d0b94f36be605b10b82ce4876bf/rainsd/inbox.go#L214)):*
<details><summary><i>Previously, britram (Brian Trammell) wrote…</i></summary><blockquote>

why is this here?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/rains/6)*
<!-- Sent from Reviewable.io -->

britram commented:
Reviewed 1 of 13 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[rainsd/inbox.go, line 188 at r2](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgLwdpRLC5746CgB50Z:-KgOLcABvSTHaBh8lJO1:br10k8i) ([raw file](https://github.com/netsec-ethz/rains/blob/6d4b139b5f1e9d0b94f36be605b10b82ce4876bf/rainsd/inbox.go#L188)):*
<details><summary><i>Previously, fehlmach wrote…</i></summary><blockquote>

To avoid spinning all the time. If there are no messages to process available then wait some time before looking again.  But we can remove it. 
</blockquote></details>

This is a select switch. The default sleep can be removed.

---

*[rainsd/inbox.go, line 201 at r2](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgLwiRCvnLVcsrZACVB:-KgORABg1OrcatAoT89B:byr4ffv) ([raw file](https://github.com/netsec-ethz/rains/blob/6d4b139b5f1e9d0b94f36be605b10b82ce4876bf/rainsd/inbox.go#L201)):*
<details><summary><i>Previously, fehlmach wrote…</i></summary><blockquote>

Done.
</blockquote></details>

see above

---

*[rainsd/inbox.go, line 214 at r2](https://reviewable.io:443/reviews/netsec-ethz/rains/6#-KgLwkeV134tdt7TEGNV:-KgORDj1p239RUgvVdde:b-34vmyd) ([raw file](https://github.com/netsec-ethz/rains/blob/6d4b139b5f1e9d0b94f36be605b10b82ce4876bf/rainsd/inbox.go#L214)):*
<details><summary><i>Previously, fehlmach wrote…</i></summary><blockquote>

Done.
</blockquote></details>

See above

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/rains/6)*
<!-- Sent from Reviewable.io -->
